cpp/src/neighbors/detail/vamana/robust_prune.cuh

```diff
@@ -0,0 +1,244 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cub/cub.cuh>
+#include <thrust/sort.h>
+
+#include "macros.cuh"
+#include "vamana_structs.cuh"
+
+namespace cuvs::neighbors::vamana::detail {
+
+// Load candidates (from query) and previous edges (from nbh_list) into registers (tmp) spanning
+// warp
+template <typename accT, typename IdxT = uint32_t, int DEG, int CANDS>
+__forceinline__ __device__ void load_to_registers(DistPair<IdxT, accT>* tmp,
+                                                  QueryCandidates<IdxT, accT>* query,
+                                                  DistPair<IdxT, accT>* nbh_list)
+{
+  int cands_per_thread = CANDS / 32;
+  for (int i = 0; i < cands_per_thread; i++) {
+    tmp[i].idx  = query->ids[cands_per_thread * threadIdx.x + i];
+    tmp[i].dist = query->dists[cands_per_thread * threadIdx.x + i];
+
+    if (cands_per_thread * threadIdx.x + i >= query->size) {
+      tmp[i].idx  = raft::upper_bound<IdxT>();
+      tmp[i].dist = raft::upper_bound<accT>();
+    }
+  }
+  int nbh_per_thread = DEG / 32;
+  for (int i = 0; i < nbh_per_thread; i++) {
+    tmp[cands_per_thread + i] = nbh_list[nbh_per_thread * threadIdx.x + i];
+  }
+}
+
+/* Combines edge and candidate lists, removes duplicates, and sorts by distance
+ * Uses CUB primitives, so needs to be templated. Called with Macros for supported sizes above */
+template <typename accT, typename IdxT, int DEG, int CANDS>
+__forceinline__ __device__ void sort_edges_and_cands(
+  DistPair<IdxT, accT>* new_nbh_list,
+  QueryCandidates<IdxT, accT>* query,
+  typename cub::BlockMergeSort<DistPair<IdxT, accT>, 32, (DEG + CANDS) / 32>::TempStorage* sort_mem)
+{
+  const int ELTS   = (DEG + CANDS) / 32;
+  using BlockSortT = cub::BlockMergeSort<DistPair<IdxT, accT>, 32, ELTS>;
+  DistPair<IdxT, accT> tmp[ELTS];
+
+  load_to_registers<accT, IdxT, DEG, CANDS>(tmp, query, new_nbh_list);
+
+  __syncthreads();
+  BlockSortT(*sort_mem).Sort(tmp, CmpDist());
+  __syncthreads();
+
+  // Mark duplicates and re-sort
+  // Copy last element over and shuffle to check for duplicate between threads
+  new_nbh_list[ELTS * threadIdx.x + (ELTS - 1)] = tmp[ELTS - 1];
+  if (tmp[ELTS - 1].idx == tmp[ELTS - 2].idx) {
+    new_nbh_list[ELTS * threadIdx.x + (ELTS - 1)].idx  = raft::upper_bound<IdxT>();
+    new_nbh_list[ELTS * threadIdx.x + (ELTS - 1)].dist = raft::upper_bound<accT>();
+  }
+  __shfl_up_sync(0xffffffff, tmp[ELTS - 1].idx, 1);
```

Comment:
I am still seeing some mentions of `__shfl_up_sync` which we might be able to replace with the corresponding raft primitives. 
@achirkin if I recall correctly, raft::shfl_up should be a wrapper around the sync primitive?

Reply:
It looks like there is only a `raft::shfl_down`. I could adjust this, but would need to re-work some of the logic and make sure it is correct / doesn't effect perf.